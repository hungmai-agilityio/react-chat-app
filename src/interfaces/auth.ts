import { ChangeEvent, FormEvent } from 'react';

export interface AuthProps {
  nameValue?: string;
  mailValue: string;
  passwordValue: string;
  confirmValue?: string;
  onNameChange?: (
    e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => void;
  onMailChange: (
    e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => void;
  onPasswordChange: (
    e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => void;
  onConfirmChange?: (
    e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
  ) => void;
  onClick: (e: FormEvent<HTMLButtonElement>) => void;
  message?: { [key: string]: string };
  authMessage?: string;
}

export interface ApiResponse<T> {
  data: T | null;
  error: unknown;
}