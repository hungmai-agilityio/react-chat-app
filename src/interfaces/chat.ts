export interface IMessage {
  id: string;
  message: string;
  sender: string;
  time_stamp: Date;
  roomId: string;
  isEdit?: boolean;
}

export interface IChat {
  id: string;
  title: string;
  avatar: string;
  members: string[];
  isGroup: boolean;
  owner?: string;
}
