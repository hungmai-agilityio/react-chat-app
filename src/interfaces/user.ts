export interface IUser {
  id: string;
  userName: string;
  email: string;
  password: string;
  created_at: Date;
  updated_at: Date;
}

export interface IProfile {
  id: string;
  user_id: string;
  avatar: string;
  phone: string;
}