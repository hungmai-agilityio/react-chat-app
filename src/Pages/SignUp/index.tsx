import React, { useState, ChangeEvent, FormEvent, useCallback } from 'react';
import { useNavigate } from 'react-router-dom';
import { v4 as uuidv4 } from 'uuid';

// Firebases
import { createUserWithEmailAndPassword } from 'firebase/auth';
import { auth } from '../../../fireBase/config';
// Interfaces
import { IUser, IProfile } from '@/interfaces';

// Services
import { addUser, addProfileUser, getUserByEmail } from '@/services';

// Utils
import { validateForm } from '@/utils';

// Constants
import { MESSAGE_API, MESSAGE_VALID, URL } from '@/constants';

// Components
import SignUp from '@/components/Form/SignUp';

const SignUpPage = () => {
  const [authMessage, setAuthMessage] = useState<string>('');
  const [name, setName] = useState<string>('');
  const [email, setEmail] = useState<string>('');
  const [password, setPassword] = useState<string>('');
  const [confirm, setConfirm] = useState<string>('');
  const [errorMessage, setErrorMessage] = useState<{ [key: string]: string }>(
    {}
  );
  const navigate = useNavigate();

  // Handle Sign up new user
  const handleSignUp = useCallback (async (e: FormEvent<HTMLButtonElement>) => {
    e.preventDefault();

    const errors = validateForm({ name, email, password, confirm });
    if (Object.keys(errors).length > 0) {
      setErrorMessage(errors);
      return;
    }

    setErrorMessage({});

    const existingUser = await getUserByEmail(email);

    if (existingUser.data) {
      setAuthMessage(MESSAGE_VALID.EMAIL_EXIST);
      return;
    }

    const userCredential = await createUserWithEmailAndPassword(
      auth,
      email,
      password
    );
    const user = userCredential.user;

    const newAccount: IUser = {
      id: user.uid,
      userName: name,
      email,
      password,
      created_at: new Date(),
      updated_at: new Date()
    };

    const account = await addUser(newAccount);

    const newProfile: IProfile = {
      id: uuidv4(),
      user_id: newAccount.id,
      avatar: '',
      phone: ''
    };

    const profile = await addProfileUser(newProfile);

    if (!account.data || !profile.data) {
      setAuthMessage(MESSAGE_API.SIGN_UP_ERROR);
      return;
    }

    setAuthMessage('');
    navigate(URL.SIGN_IN);
  }, [confirm, email, name, navigate, password]);

  // Handle change input value
  const handleInputChange =
    (setter: React.Dispatch<React.SetStateAction<string>>) =>
    (e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
      setter(e.target.value);
    };

  return (
    <SignUp
      nameValue={name}
      mailValue={email}
      passwordValue={password}
      confirmValue={confirm}
      onNameChange={handleInputChange(setName)}
      onMailChange={handleInputChange(setEmail)}
      onPasswordChange={handleInputChange(setPassword)}
      onConfirmChange={handleInputChange(setConfirm)}
      onClick={handleSignUp}
      message={errorMessage}
      authMessage={authMessage}
    />
  );
};

export default SignUpPage;
