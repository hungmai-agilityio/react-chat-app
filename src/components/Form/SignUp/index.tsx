import { useCallback, useState } from 'react';
import { useNavigate } from 'react-router-dom';

// Font Awesome
import { faEye, faEyeSlash } from '@fortawesome/free-solid-svg-icons';

// Constants
import { SIZE, TYPE, URL } from '@/constants';

// Interface
import { AuthProps } from '@/interfaces';

// Components
import Button from '@/components/Button';
import Form from '@/components/Form';
import Input from '@/components/Input';

const SignUp = ({
  nameValue,
  mailValue,
  passwordValue,
  confirmValue,
  onNameChange,
  onMailChange,
  onPasswordChange,
  onConfirmChange,
  onClick,
  message,
  authMessage,
  isDisabled
}: AuthProps) => {
  const [showPassword, setShowPassword] = useState<boolean>(false);
  const [showConfirmPassword, setShowConfirmPassword] =
    useState<boolean>(false);
  const navigate = useNavigate();

  const toggleShowPassword = useCallback(() => {
    setShowPassword(!showPassword);
  }, [showPassword]);

  const toggleShowConfirmPassword = useCallback(() => {
    setShowConfirmPassword(!showConfirmPassword);
  }, [showConfirmPassword]);

  const handleMoveToSignIn = () => {
    navigate(URL.SIGN_IN);
  };

  return (
    <Form title="Registration">
      <form className="mt-11">
        <Input
          label="Name"
          placeholder="Username"
          htmlFor="name"
          value={nameValue}
          onChange={onNameChange!}
          message={message?.name}
        />
        <Input
          label="Email"
          placeholder="username@mail.com"
          htmlFor="email"
          value={mailValue}
          onChange={onMailChange}
          type="email"
          message={message?.email}
        />
        <Input
          label="Password"
          placeholder="•••••••••••••"
          htmlFor="password"
          value={passwordValue}
          onChange={onPasswordChange}
          icon={showPassword ? faEye : faEyeSlash}
          type={showPassword ? 'text' : 'password'}
          onClick={toggleShowPassword}
          message={message?.password}
        />
        <Input
          label="Confirm Password"
          placeholder="•••••••••••••"
          htmlFor="confirm"
          value={confirmValue}
          onChange={onConfirmChange!}
          icon={showConfirmPassword ? faEye : faEyeSlash}
          type={showConfirmPassword ? TYPE.TEXT : TYPE.PASSWORD}
          onClick={toggleShowConfirmPassword}
          message={message?.confirm}
        />
        <div className="mt-8">
          <Button
            name="Register"
            variant={TYPE.PRIMARY}
            onClick={onClick!}
            size={SIZE.LARGE}
            disabled={isDisabled}
          />
          {authMessage && (
            <p className="mt-3 text-center text-red-500">{authMessage}</p>
          )}
          <p className="text-sm mt-10">
            Already a user?{' '}
            <span
              className="text-dark underline cursor-pointer"
              onClick={handleMoveToSignIn}
            >
              Login now
            </span>
          </p>
        </div>
      </form>
    </Form>
  );
};

export default SignUp;
