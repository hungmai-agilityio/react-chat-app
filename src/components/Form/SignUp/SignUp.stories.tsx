import { Meta, StoryObj } from '@storybook/react';
import { withRouter } from 'storybook-addon-remix-react-router';

// Components
import SignUp from '@/components/Form/SignUp';

const meta: Meta<typeof SignUp> = {
  component: SignUp,
  tags: ['autodocs'],
  argTypes: {
    mailValue: {
      description: 'Set the value for the mail input field'
    },
    passwordValue: {
      description: 'Set the value for the password input field'
    },
    confirmValue: {
      description: 'Set the value for the confirm input field'
    },
    nameValue: {
      description: 'Set the value for the name input field'
    },
    onMailChange: {
      description: 'Handle change value mail field'
    },
    onPasswordChange: {
      description: 'Handle change value password field'
    },
    onConfirmChange: {
      description: 'Handle change value confirm field'
    },
    onNameChange: {
      description: 'Handle change value name field'
    },
    onClick: {
      description: 'Handle event when click '
    },
    message: {
      description: 'Show error messages according to the input field '
    },
    authMessage: {
      description: 'Show api error messages'
    }
  },
  decorators: [withRouter]
};

export default meta;

type Story = StoryObj<typeof SignUp>;

export const Primary: Story = {
  args: {
    nameValue: 'Name',
    mailValue: 'username@mail.com',
    passwordValue: 'Password123',
    confirmValue: 'Password123',
    onNameChange: () => {},
    onMailChange: () => {},
    onPasswordChange: () => {},
    onConfirmChange: () => {},
    onClick: () => {}
  }
};

export const MessageError: Story = {
  args: {
    nameValue: '',
    mailValue: '',
    passwordValue: '',
    confirmValue: '',
    onNameChange: () => {},
    onMailChange: () => {},
    onPasswordChange: () => {},
    onConfirmChange: () => {},
    onClick: () => {},
    message: {
      name: 'Name is required',
      email: 'Email is required',
      password: 'Password is required',
      confirm: 'Confirm Password is required'
    }
  }
};
