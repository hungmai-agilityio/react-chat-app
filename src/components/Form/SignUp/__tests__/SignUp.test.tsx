import { fireEvent, render, screen } from '@testing-library/react';
import SignUp from '@/components/Form/SignUp';
import { AuthProps } from '@/interfaces';
import { MemoryRouter } from 'react-router-dom';

describe('SignUp Component', () => {
  const onClickMock = jest.fn((e) => e.preventDefault());
  const defaultProps: AuthProps = {
    nameValue: '',
    mailValue: '',
    passwordValue: '',
    confirmValue: '',
    onNameChange: onClickMock,
    onMailChange: onClickMock,
    onPasswordChange: onClickMock,
    onConfirmChange: onClickMock,
    onClick: onClickMock,
    message: {
      name: '',
      email: '',
      password: '',
      confirm: ''
    },
    authMessage: ''
  };

  const renderWithRouter = (ui: React.ReactElement) => {
    return render(ui, { wrapper: MemoryRouter });
  };

  it('renders SignUp component with default empty props and displays message on button click', () => {
    renderWithRouter(<SignUp {...defaultProps} />);

    expect(screen.getByText('Registration')).toBeInTheDocument();
    expect(screen.getByLabelText('Name')).toBeInTheDocument();
    expect(screen.getByLabelText('Email')).toBeInTheDocument();
    expect(screen.getByLabelText('Password')).toBeInTheDocument();
    expect(screen.getByLabelText('Confirm Password')).toBeInTheDocument();

    const registerButton = screen.getByText('Register');
    fireEvent.click(registerButton!);

    expect(defaultProps.onClick).toHaveBeenCalled();

    defaultProps.message = {
      name: 'Name is required',
      email: 'Email is required',
      password: 'Password is required',
      confirm: 'Confirm is required'
    };

    renderWithRouter(<SignUp {...defaultProps} />);
    expect(screen.getByText('Name is required')).toBeInTheDocument();
    expect(screen.getByText('Email is required')).toBeInTheDocument();
    expect(screen.getByText('Password is required')).toBeInTheDocument();
    expect(screen.getByText('Confirm is required')).toBeInTheDocument();
  });

  it('renders SignUp component with provided values and toggles password visibility on icon click', () => {
    const { container } = renderWithRouter(
      <SignUp
        {...defaultProps}
        nameValue="John Doe"
        mailValue="test@mail.com"
        passwordValue="password123"
        confirmValue="password123"
      />
    );

    expect(screen.getByLabelText('Name')).toHaveValue('John Doe');
    expect(screen.getByLabelText('Email')).toHaveValue('test@mail.com');
    expect(screen.getByLabelText('Password')).toHaveValue('password123');
    expect(screen.getByLabelText('Confirm Password')).toHaveValue(
      'password123'
    );

    const passwordInput = screen.getByLabelText('Password');
    expect(passwordInput).toHaveAttribute('type', 'password');

    const confirmInput = screen.getByLabelText('Confirm Password');
    expect(confirmInput).toHaveAttribute('type', 'password');

    const passwordElement = container.querySelector('#password');
    const passwordIcon = passwordElement?.nextElementSibling;
    expect(passwordIcon).toBeInTheDocument();

    fireEvent.click(passwordIcon!);

    expect(passwordInput).toHaveAttribute('type', 'text');
    expect(confirmInput).toHaveAttribute('type', 'password');

    const confirmElement = container.querySelector('#confirm');
    const confirmIcon = confirmElement?.nextElementSibling;
    expect(passwordIcon).toBeInTheDocument();

    fireEvent.click(confirmIcon!);

    expect(passwordInput).toHaveAttribute('type', 'text');
    expect(confirmInput).toHaveAttribute('type', 'text');
  });

  it('should render match to snapshot', () => {
    const { container } = renderWithRouter(<SignUp {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });
});
