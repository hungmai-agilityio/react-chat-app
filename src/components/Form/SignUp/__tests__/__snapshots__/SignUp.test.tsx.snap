// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp Component should render match to snapshot 1`] = `
<div>
  <div
    class="h-screen flex bg-auth bg-no-repeat bg-cover overflow-hidden"
  >
    <div
      class="w-form m-auto bg-tertiary p-8 rounded-2xl"
    >
      <img
        alt="Logo"
        class="block my-0 mx-auto w-32 "
        src="assets/images/chat-app-logo.png"
      />
      <p
        class="mt-4 text-center text-primary font-semibold text-xl"
      >
        Registration
      </p>
      <form
        class="mt-11"
      >
        <div
          class="mb-3"
        >
          <label
            class="font-bold leading-2-xs capitalize"
            for="name"
          >
            Name
          </label>
          <div
            class="relative"
          >
            <input
              class="w-full mt-1.5 outline-none p-3 focus:border-primary h-11 border border-gray-400 rounded-md"
              id="name"
              placeholder="Username"
              type="text"
              value=""
            />
          </div>
          <span
            class="text-red-500 mr-1 visible text-xs"
          >
            Name is required
          </span>
        </div>
        <div
          class="mb-3"
        >
          <label
            class="font-bold leading-2-xs capitalize"
            for="email"
          >
            Email
          </label>
          <div
            class="relative"
          >
            <input
              class="w-full mt-1.5 outline-none p-3 focus:border-primary h-11 border border-gray-400 rounded-md"
              id="email"
              placeholder="username@mail.com"
              type="email"
              value=""
            />
          </div>
          <span
            class="text-red-500 mr-1 visible text-xs"
          >
            Email is required
          </span>
        </div>
        <div
          class="mb-3"
        >
          <label
            class="font-bold leading-2-xs capitalize"
            for="password"
          >
            Password
          </label>
          <div
            class="relative"
          >
            <input
              class="w-full mt-1.5 outline-none p-3 focus:border-primary h-11 border border-gray-400 rounded-md"
              id="password"
              placeholder="•••••••••••••"
              type="password"
              value=""
            />
            <div
              class="flex items-center justify-center cursor-pointer text-gray-400 h-10 w-10 rounded-full hover:text-dark hover:bg-tertiary absolute top-2 right-2"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-eye-slash "
                data-icon="eye-slash"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 640 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M38.8 5.1C28.4-3.1 13.3-1.2 5.1 9.2S-1.2 34.7 9.2 42.9l592 464c10.4 8.2 25.5 6.3 33.7-4.1s6.3-25.5-4.1-33.7L525.6 386.7c39.6-40.6 66.4-86.1 79.9-118.4c3.3-7.9 3.3-16.7 0-24.6c-14.9-35.7-46.2-87.7-93-131.1C465.5 68.8 400.8 32 320 32c-68.2 0-125 26.3-169.3 60.8L38.8 5.1zM223.1 149.5C248.6 126.2 282.7 112 320 112c79.5 0 144 64.5 144 144c0 24.9-6.3 48.3-17.4 68.7L408 294.5c8.4-19.3 10.6-41.4 4.8-63.3c-11.1-41.5-47.8-69.4-88.6-71.1c-5.8-.2-9.2 6.1-7.4 11.7c2.1 6.4 3.3 13.2 3.3 20.3c0 10.2-2.4 19.8-6.6 28.3l-90.3-70.8zM373 389.9c-16.4 6.5-34.3 10.1-53 10.1c-79.5 0-144-64.5-144-144c0-6.9 .5-13.6 1.4-20.2L83.1 161.5C60.3 191.2 44 220.8 34.5 243.7c-3.3 7.9-3.3 16.7 0 24.6c14.9 35.7 46.2 87.7 93 131.1C174.5 443.2 239.2 480 320 480c47.8 0 89.9-12.9 126.2-32.5L373 389.9z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </div>
          <span
            class="text-red-500 mr-1 visible text-xs"
          >
            Password is required
          </span>
        </div>
        <div
          class="mb-3"
        >
          <label
            class="font-bold leading-2-xs capitalize"
            for="confirm"
          >
            Confirm Password
          </label>
          <div
            class="relative"
          >
            <input
              class="w-full mt-1.5 outline-none p-3 focus:border-primary h-11 border border-gray-400 rounded-md"
              id="confirm"
              placeholder="•••••••••••••"
              type="password"
              value=""
            />
            <div
              class="flex items-center justify-center cursor-pointer text-gray-400 h-10 w-10 rounded-full hover:text-dark hover:bg-tertiary absolute top-2 right-2"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-eye-slash "
                data-icon="eye-slash"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 640 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M38.8 5.1C28.4-3.1 13.3-1.2 5.1 9.2S-1.2 34.7 9.2 42.9l592 464c10.4 8.2 25.5 6.3 33.7-4.1s6.3-25.5-4.1-33.7L525.6 386.7c39.6-40.6 66.4-86.1 79.9-118.4c3.3-7.9 3.3-16.7 0-24.6c-14.9-35.7-46.2-87.7-93-131.1C465.5 68.8 400.8 32 320 32c-68.2 0-125 26.3-169.3 60.8L38.8 5.1zM223.1 149.5C248.6 126.2 282.7 112 320 112c79.5 0 144 64.5 144 144c0 24.9-6.3 48.3-17.4 68.7L408 294.5c8.4-19.3 10.6-41.4 4.8-63.3c-11.1-41.5-47.8-69.4-88.6-71.1c-5.8-.2-9.2 6.1-7.4 11.7c2.1 6.4 3.3 13.2 3.3 20.3c0 10.2-2.4 19.8-6.6 28.3l-90.3-70.8zM373 389.9c-16.4 6.5-34.3 10.1-53 10.1c-79.5 0-144-64.5-144-144c0-6.9 .5-13.6 1.4-20.2L83.1 161.5C60.3 191.2 44 220.8 34.5 243.7c-3.3 7.9-3.3 16.7 0 24.6c14.9 35.7 46.2 87.7 93 131.1C174.5 443.2 239.2 480 320 480c47.8 0 89.9-12.9 126.2-32.5L373 389.9z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </div>
          <span
            class="text-red-500 mr-1 visible text-xs"
          >
            Confirm is required
          </span>
        </div>
        <div
          class="mt-8"
        >
          <button
            class="font-semibold hover:brightness-90 w-full h-10 bg-primary text-white"
          >
            Register
          </button>
          <p
            class="text-sm mt-10"
          >
            Already a user?
             
            <span
              class="text-dark underline cursor-pointer"
            >
              Login now
            </span>
          </p>
        </div>
      </form>
    </div>
  </div>
</div>
`;
